File: go/jsonrpc3/peer.go
package jsonrpc3

import (
	"context"
	"fmt"
	"io"
	"os"
	"sync"
	"sync/atomic"
	"time"
)

// Peer represents a bidirectional JSON-RPC 3.0 peer over a stream-based transport.
// It supports full bidirectional communication where both peers can initiate method
// calls at any time. Works with any io.Reader/io.Writer pair (pipes, sockets, etc.).
type Peer struct {
	reader      io.Reader
	writer      io.Writer
	session     *Session
	handler     *Handler
	rootObject  Object
	contentType string

	// Request tracking for concurrent requests
	pendingReqs sync.Map // id (any) -> chan *Response
	nextID      atomic.Int64

	// Reference ID generation
	refPrefix  string // Random connection-specific prefix for refs
	refCounter atomic.Int64

	// Message channels
	writeChan chan MessageSetConvertible // Messages to encode and send
	closeChan chan struct{}
	closeOnce sync.Once

	// Context for cancelling operations
	ctx    context.Context
	cancel context.CancelFunc

	// Error tracking
	connErr error
	errMu   sync.RWMutex
}

// NewPeer creates a new bidirectional JSON-RPC 3.0 peer over a stream-based transport.
// The rootObject handles incoming method calls from the remote peer.
// Default encoding is MimeTypeJSON.
//
// Options:
//   - WithContentType(contentType) - specify encoding format
//
// Example:
//
//	// Using pipes
//	r1, w1 := io.Pipe()
//	r2, w2 := io.Pipe()
//	peer1 := NewPeer(r1, w2, serverRoot)
//	peer2 := NewPeer(r2, w1, clientRoot)
func NewPeer(reader io.Reader, writer io.Writer, rootObject Object, opts ...ClientOption) (*Peer, error) {
	// Apply options with defaults
	options := &clientOptions{
		contentType: MimeTypeJSON,
	}
	for _, opt := range opts {
		opt(options)
	}

	// Create context for lifecycle management
	ctx, cancel := context.WithCancel(context.Background())

	// Create session and handler
	session := NewSession()
	mimeTypes := []string{options.contentType}
	handler := NewHandler(session, rootObject, mimeTypes)

	// Generate random connection ID prefix
	refPrefix := generateConnID()

	peer := &Peer{
		reader:      reader,
		writer:      writer,
		session:     session,
		handler:     handler,
		rootObject:  rootObject,
		contentType: options.contentType,
		refPrefix:   refPrefix,
		writeChan:   make(chan MessageSetConvertible, 100),
		closeChan:   make(chan struct{}),
		ctx:         ctx,
		cancel:      cancel,
	}

	// Start read and write loops
	go peer.readLoop()
	go peer.writeLoop()

	return peer, nil
}

// Stdio creates a new Peer using standard input and output streams.
// The rootObject handles incoming method calls from the remote peer.
// Default encoding is MimeTypeJSON.
func Stdio(rootObject Object, opts ...ClientOption) (*Peer, error) {
	if mimetype := os.Getenv("JSONRPC_CONTENT_TYPE"); mimetype != "" {
		switch mimetype {
		case MimeTypeJSON:
			opts = append(opts, WithJSON())
			// Default, do nothing
		case MimeTypeCBOR:
			opts = append(opts, WithCBOR())
		case MimeTypeCBORCompact:
			opts = append(opts, WithCompactCBOR())
		default:
			return nil, fmt.Errorf("unsupported JSON-RPC content type: %s", mimetype)
		}
	} else {
		opts = append([]ClientOption{WithJSON()}, opts...)
	}
	return NewPeer(os.Stdin, os.Stdout, rootObject, opts...)
}

// Call invokes a method on the remote peer and waits for the response.
// Use the ToRef option to call a method on a remote object reference.
func (p *Peer) Call(method string, params any, result any, opts ...CallOption) error {
	// Check if connection is closed
	if err := p.getError(); err != nil {
		return err
	}

	// Apply options
	var options callOptions
	for _, opt := range opts {
		opt.apply(&options)
	}

	// Generate request ID
	// Use float64 because JSON unmarshaling converts numbers to float64
	id := float64(p.nextID.Add(1))

	// Create request
	req, err := NewRequestWithFormat(method, params, id, p.contentType)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	if options.ref != nil {
		req.Ref = options.ref.Ref
	}

	// Create response channel
	respChan := make(chan *Response, 1)
	p.pendingReqs.Store(id, respChan)
	defer p.pendingReqs.Delete(id)

	// Send request via write channel (will be encoded in writeLoop)
	select {
	case p.writeChan <- req:
		// Successfully queued for sending
	case <-p.closeChan:
		return fmt.Errorf("connection closed")
	}

	// Wait for response with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	select {
	case resp := <-respChan:
		if resp.Error != nil {
			return resp.Error
		}

		// Decode result if provided
		if result != nil && resp.Result != nil {
			params := NewParamsWithFormat(resp.Result, p.contentType)
			if err := params.Decode(result); err != nil {
				return fmt.Errorf("failed to decode result: %w", err)
			}
		}
		return nil

	case <-ctx.Done():
		return ctx.Err()

	case <-p.closeChan:
		return fmt.Errorf("connection closed")
	}
}

// Notify sends a notification (no response expected).
// Use the ToRef option to send a notification to a remote object reference.
func (p *Peer) Notify(method string, params any, opts ...CallOption) error {
	// Check if connection is closed
	if err := p.getError(); err != nil {
		return err
	}

	// Apply options
	var options callOptions
	for _, opt := range opts {
		opt.apply(&options)
	}

	// Create notification (nil ID)
	req, err := NewRequestWithFormat(method, params, nil, p.contentType)
	if err != nil {
		return fmt.Errorf("failed to create notification: %w", err)
	}

	if options.ref != nil {
		req.Ref = options.ref.Ref
	}

	// Send notification via write channel (will be encoded in writeLoop)
	select {
	case p.writeChan <- req:
		return nil
	case <-p.closeChan:
		return fmt.Errorf("connection closed")
	}
}

// RegisterObject registers a local object that the remote peer can call.
// If ref is empty, a reference ID is auto-generated using the connection's prefix and counter.
// Returns the reference ID that was used (either the provided one or the generated one).
func (p *Peer) RegisterObject(ref string, obj Object) string {
	if ref == "" {
		counter := p.refCounter.Add(1)
		ref = fmt.Sprintf("%s-%d", p.refPrefix, counter)
	}
	p.handler.session.AddLocalRef(ref, obj)
	return ref
}

// UnregisterObject removes a registered local object.
func (p *Peer) UnregisterObject(ref string) {
	p.handler.session.RemoveLocalRef(ref)
}

// CallBatch sends a batch of requests and returns the results.
// This is the low-level batch interface required by the Caller interface.
// For a more convenient builder-based API, use the ExecuteBatch function.
//
// Example:
//
//	requests := []BatchRequest{
//	    {Method: "method1", Params: params1},
//	    {Method: "method2", Params: params2},
//	}
//	results, err := peer.CallBatch(requests)
func (p *Peer) CallBatch(batchReqs []BatchRequest) (*BatchResults, error) {
	// Check if connection is closed
	if err := p.getError(); err != nil {
		return nil, err
	}

	// Check if batch is empty
	if len(batchReqs) == 0 {
		return &BatchResults{responses: []*Response{}}, nil
	}

	// Convert BatchRequest to Request objects
	requests := make([]*Request, len(batchReqs))
	for i, breq := range batchReqs {
		var id any
		if !breq.IsNotification {
			id = float64(p.nextID.Add(1))
		}

		req, err := NewRequest(breq.Method, breq.Params, id)
		if err != nil {
			return nil, fmt.Errorf("failed to create request %d: %w", i, err)
		}

		if breq.Ref != "" {
			req.Ref = breq.Ref
		}

		requests[i] = req
	}

	// Create response channels for each request with an ID
	responseChans := make(map[any]chan *Response)
	for _, req := range requests {
		if req.ID != nil {
			respChan := make(chan *Response, 1)
			responseChans[req.ID] = respChan
			p.pendingReqs.Store(req.ID, respChan)
		}
	}

	// Clean up pending requests when done
	defer func() {
		for id := range responseChans {
			p.pendingReqs.Delete(id)
		}
	}()

	// Create batch message
	msgSet := &MessageSet{
		Messages: make([]Message, len(requests)),
		IsBatch:  true,
	}
	for i, req := range requests {
		msgSet.Messages[i] = Message{
			JSONRPC: req.JSONRPC,
			Method:  req.Method,
			Params:  req.Params,
			ID:      req.ID,
			Ref:     req.Ref,
		}
	}

	// Send batch via write channel
	select {
	case p.writeChan <- msgSet:
		// Successfully queued for sending
	case <-p.closeChan:
		return nil, fmt.Errorf("connection closed")
	}

	// Wait for all responses with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	responses := make([]*Response, len(requests))
	responsesReceived := 0

	// Collect responses
	for i, req := range requests {
		// Skip notifications (they don't have responses)
		if req.ID == nil {
			responses[i] = nil
			responsesReceived++
			continue
		}

		respChan, ok := responseChans[req.ID]
		if !ok {
			return nil, fmt.Errorf("missing response channel for request %v", req.ID)
		}

		select {
		case resp := <-respChan:
			responses[i] = resp
			responsesReceived++

		case <-ctx.Done():
			return nil, fmt.Errorf("timeout waiting for response %d: %w", i, ctx.Err())

		case <-p.closeChan:
			return nil, fmt.Errorf("connection closed while waiting for responses")
		}
	}

	return &BatchResults{responses: responses}, nil
}

// GetSession returns the peer's session.
func (p *Peer) GetSession() *Session {
	return p.session
}

// Close closes the peer connection gracefully.
func (p *Peer) Close() error {
	var err error
	p.closeOnce.Do(func() {
		// Cancel context to stop any pending operations
		p.cancel()

		close(p.closeChan)

		// Dispose all refs
		p.session.DisposeAll()

		// Wake up all pending requests
		p.pendingReqs.Range(func(key, value any) bool {
			if ch, ok := value.(chan *Response); ok {
				close(ch)
			}
			p.pendingReqs.Delete(key)
			return true
		})
	})
	return err
}

// readLoop continuously reads messages from the reader.
func (p *Peer) readLoop() {
	defer p.Close()

	codec := GetCodec(p.contentType)
	decoder := codec.NewMessageDecoder(p.reader)

	for {
		select {
		case <-p.closeChan:
			return
		default:
		}

		// Decode message set from stream (could be single message or batch)
		msgSet, err := decoder.Decode()
		if err != nil {
			if err != io.EOF {
				p.setError(fmt.Errorf("read error: %w", err))
			}
			return
		}

		// Check if all messages are requests or all are responses
		allRequests := true
		allResponses := true
		for i := range msgSet.Messages {
			if !msgSet.Messages[i].IsRequest() {
				allRequests = false
			}
			if !msgSet.Messages[i].IsResponse() {
				allResponses = false
			}
		}

		// Handle batch requests as a unit (preserves batch-local references)
		if allRequests && msgSet.IsBatch {
			go p.handleIncomingBatch(&msgSet)
		} else if allRequests {
			// Single request
			msg := &msgSet.Messages[0]
			go p.handleIncomingRequest(msg)
		} else if allResponses {
			// Process responses individually (no batch context needed)
			for i := range msgSet.Messages {
				msg := &msgSet.Messages[i]
				p.handleIncomingResponse(msg)
			}
		} else {
			// Mixed request/response batches are invalid
			// Send error responses for all requests in the batch
			go p.handleInvalidBatch(&msgSet)
		}
	}
}

// handleIncomingRequest processes an incoming request from the remote peer.
func (p *Peer) handleIncomingRequest(msg *Message) {
	req := msg.ToRequest()
	if req == nil {
		return
	}

	// Handle request via handler
	resp := p.handler.HandleRequest(req)

	// Send response if not a notification (will be encoded in writeLoop)
	if resp != nil {
		select {
		case p.writeChan <- resp:
		case <-p.closeChan:
		}
	}
}

// sendErrorForBatch sends error responses for all non-notification requests in a MessageSet.
func (p *Peer) sendErrorForBatch(msgSet *MessageSet, rpcErr *Error) {
	var responses BatchResponse

	// Create error responses for all requests (ignore responses)
	for i := range msgSet.Messages {
		msg := &msgSet.Messages[i]
		if msg.IsRequest() {
			req := msg.ToRequest()
			if req != nil && !req.IsNotification() {
				errResp := NewErrorResponse(req.ID, rpcErr, Version30)
				responses = append(responses, *errResp)
			}
		}
	}

	// Send error responses if any (will be encoded in writeLoop)
	if len(responses) > 0 {
		select {
		case p.writeChan <- responses:
		case <-p.closeChan:
		}
	}
}

// handleIncomingBatch processes a batch of incoming requests from the remote peer.
// Batch requests must be processed together to support batch-local references (\0, \1, etc.)
func (p *Peer) handleIncomingBatch(msgSet *MessageSet) {
	// Convert MessageSet to Batch
	batch, err := msgSet.ToBatch()
	if err != nil {
		// Failed to convert - send error responses for all requests
		p.sendErrorForBatch(msgSet, NewInternalError("failed to parse batch: "+err.Error()))
		return
	}

	// Handle batch via handler (preserves batch-local references)
	responses := p.handler.HandleBatch(batch)

	// Send batch response as a unit (will be encoded in writeLoop)
	if len(responses) > 0 {
		select {
		case p.writeChan <- responses:
		case <-p.closeChan:
		}
	}
}

// handleInvalidBatch handles a batch with mixed requests and responses.
// Sends error responses for all requests in the batch.
func (p *Peer) handleInvalidBatch(msgSet *MessageSet) {
	p.sendErrorForBatch(msgSet, NewInvalidRequestError("mixed requests and responses in batch"))
}

// handleIncomingResponse processes an incoming response to our request.
func (p *Peer) handleIncomingResponse(msg *Message) {
	resp := msg.ToResponse()
	if resp == nil {
		return
	}

	// Find pending request
	if value, ok := p.pendingReqs.Load(resp.ID); ok {
		if ch, ok := value.(chan *Response); ok {
			select {
			case ch <- resp:
			default:
				// Channel full or closed, ignore
			}
		}
	}
}

// writeLoop continuously sends queued messages to the writer.
func (p *Peer) writeLoop() {
	codec := GetCodec(p.contentType)

	for {
		select {
		case item := <-p.writeChan:
			// Convert to MessageSet
			msgSet := item.ToMessageSet()

			// Encode the message set
			data, err := codec.MarshalMessages(msgSet)
			if err != nil {
				p.notifyPendingRequest(item, fmt.Errorf("marshal error: %w", err))
				p.setError(fmt.Errorf("marshal error: %w", err))
				return
			}

			// Write to stream with newline for NDJSON compatibility
			if _, err := p.writer.Write(data); err != nil {
				p.notifyPendingRequest(item, fmt.Errorf("write error: %w", err))
				p.setError(fmt.Errorf("write error: %w", err))
				return
			}
			// Add newline after each JSON message (NDJSON format)
			if _, err := p.writer.Write([]byte{'\n'}); err != nil {
				p.notifyPendingRequest(item, fmt.Errorf("write error: %w", err))
				p.setError(fmt.Errorf("write error: %w", err))
				return
			}

		case <-p.closeChan:
			return
		}
	}
}

// notifyPendingRequest checks if the item is a Request with a pending waiter,
// and if so, sends a synthetic error response to unblock the caller.
func (p *Peer) notifyPendingRequest(item any, err error) {
	// Check if this is a Request (not a Response or BatchResponse)
	req, ok := item.(*Request)
	if !ok {
		return
	}

	// Only handle requests with IDs (not notifications)
	if req.ID == nil {
		return
	}

	// Check if there's a pending waiter for this request
	if value, ok := p.pendingReqs.Load(req.ID); ok {
		if respChan, ok := value.(chan *Response); ok {
			// Create synthetic error response
			syntheticResp := &Response{
				JSONRPC: req.JSONRPC,
				Error:   NewInternalError(fmt.Sprintf("Failed to send request: %v", err)),
				ID:      req.ID,
			}

			// Try to send the error response to the waiting channel
			select {
			case respChan <- syntheticResp:
				// Successfully notified the waiter
			default:
				// Channel full or closed, ignore
			}
		}
	}
}

// setError stores a connection error.
func (p *Peer) setError(err error) {
	p.errMu.Lock()
	defer p.errMu.Unlock()
	if p.connErr == nil {
		p.connErr = err
	}
}

// getError retrieves any connection error.
func (p *Peer) getError() error {
	p.errMu.RLock()
	defer p.errMu.RUnlock()
	return p.connErr
}

File: ts/src/peer.ts
/**
 * Peer represents a bidirectional JSON-RPC 3.0 peer over a stream-based transport.
 * Both peers can initiate method calls at any time.
 */

import { nanoid } from "nanoid";
import { Session, type RpcObject } from "./session.ts";
import type { Request, Response, Message, MessageSet, Batch, BatchResponse, ReferenceType, RequestId } from "./types.ts";
import { newRequest, isReference, toBatch, batchResponseToMessageSet, Reference } from "./types.ts";
import { Handler } from "./handler.ts";
import { getCodec, MimeTypeJSON, type Codec } from "./encoding.ts";
import { RpcError } from "./error.ts";
import { JsonStreamDecoder } from "./stream-decoder.ts";

export interface PeerOptions {
  mimeType?: string;
  session?: Session;
}

interface PendingRequest {
  resolve: (value: unknown) => void;
  reject: (error: Error) => void;
}

/**
 * Promote ReferenceType objects to Reference class instances
 * This allows calling methods directly on returned references
 */
function promoteReferences(value: unknown): unknown {
  if (value === null || value === undefined) {
    return value;
  }

  // If it's a reference, promote to Reference class
  if (isReference(value)) {
    return new Reference(value);
  }

  // Recursively handle arrays
  if (Array.isArray(value)) {
    return value.map(item => promoteReferences(item));
  }

  // Recursively handle plain objects
  if (typeof value === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, val] of Object.entries(value)) {
      result[key] = promoteReferences(val);
    }
    return result;
  }

  return value;
}

/**
 * Peer for bidirectional JSON-RPC 3.0 communication over streams
 */
export class Peer {
  private codec: Codec;
  private session: Session;
  private handler: Handler;
  private reader: ReadableStreamDefaultReader<Uint8Array>;
  private writer: WritableStreamDefaultWriter<Uint8Array>;

  // Request tracking
  private pendingRequests = new Map<RequestId, PendingRequest>();
  private nextId = 1;

  // Reference ID generation
  private refPrefix: string;
  private refCounter = 0;

  // Lifecycle management
  private abortController: AbortController;
  private readLoopPromise: Promise<void>;
  private closed = false;
  private connError?: Error;

  // Write mutex to ensure serialized writes
  private writeMutex = Promise.resolve();

  // Stream decoder for handling partial JSON chunks
  private streamDecoder = new JsonStreamDecoder();

  constructor(
    readable: ReadableStream<Uint8Array>,
    writable: WritableStream<Uint8Array>,
    rootObject: RpcObject,
    options: PeerOptions = {},
  ) {
    const mimeType = options.mimeType ?? MimeTypeJSON;
    this.codec = getCodec(mimeType);

    this.session = options.session ?? new Session();
    this.handler = new Handler(this.session, rootObject);

    this.reader = readable.getReader();
    this.writer = writable.getWriter();

    this.refPrefix = nanoid();
    this.abortController = new AbortController();

    // Start read loop
    this.readLoopPromise = this.readLoop();
  }

  /**
   * Call a method on the remote peer and wait for response
   */
  async call(method: string, params?: unknown, ref?: string | Reference): Promise<unknown> {
    if (this.closed || this.connError) {
      throw this.connError ?? new Error("Connection closed");
    }

    const id = this.nextId++;
    const refString = typeof ref === "string" ? ref : ref?.$ref;
    const req = newRequest(method, params, id, refString);

    // Create promise for response
    const responsePromise = new Promise<unknown>((resolve, reject) => {
      this.pendingRequests.set(id, { resolve, reject });
    });

    // Send request
    const msgSet: MessageSet = {
      messages: [req],
      isBatch: false,
    };

    await this.writeMessage(msgSet);

    return responsePromise;
  }

  /**
   * Send a notification (no response expected)
   */
  async notify(method: string, params?: unknown, ref?: string | Reference): Promise<void> {
    if (this.closed || this.connError) {
      throw this.connError ?? new Error("Connection closed");
    }

    const refString = typeof ref === "string" ? ref : ref?.$ref;
    const req = newRequest(method, params, undefined, refString);

    const msgSet: MessageSet = {
      messages: [req],
      isBatch: false,
    };

    await this.writeMessage(msgSet);
  }

  /**
   * Send a batch of requests
   */
  async batch(requests: Request[]): Promise<BatchResponse> {
    if (this.closed || this.connError) {
      throw this.connError ?? new Error("Connection closed");
    }

    const msgSet: MessageSet = {
      messages: requests,
      isBatch: true,
    };

    // Create promises for all requests with IDs
    const responsePromises: Array<Promise<Response>> = [];

    for (const req of requests) {
      if (req.id !== undefined) {
        const promise = new Promise<Response>((resolve, reject) => {
          this.pendingRequests.set(req.id!, {
            resolve: (value) => resolve(value as Response),
            reject,
          });
        });
        responsePromises.push(promise);
      }
    }

    await this.writeMessage(msgSet);

    const responses = await Promise.all(responsePromises);
    return responses;
  }

  /**
   * Register a local object that the remote peer can call
   */
  registerObject(obj: RpcObject, ref?: string): Reference {
    if (!ref) {
      this.refCounter++;
      ref = `${this.refPrefix}-${this.refCounter}`;
    }
    this.session.addLocalRef(ref, obj);
    return new Reference(ref);
  }

  /**
   * Unregister a local object
   */
  unregisterObject(ref: string | ReferenceType): boolean {
    const refString = typeof ref === "string" ? ref : ref.$ref;
    return this.session.removeLocalRef(refString);
  }

  /**
   * Get the session
   */
  getSession(): Session {
    return this.session;
  }

  /**
   * Close the peer connection
   */
  close(): void {
    if (this.closed) {
      return;
    }

    this.closed = true;
    this.abortController.abort();

    // Reject all pending requests
    for (const [id, pending] of this.pendingRequests.entries()) {
      pending.reject(new Error("Connection closed"));
      this.pendingRequests.delete(id);
    }

    // Dispose all references
    this.session.disposeAll();

    // Close streams
    this.reader.cancel().catch(() => {});
    this.writer.close().catch(() => {});
  }

  /**
   * Wait for the peer to close
   */
  async wait(): Promise<void> {
    return this.readLoopPromise;
  }

  /**
   * Write a message to the stream (with mutex for serialization)
   */
  private async writeMessage(msgSet: MessageSet): Promise<void> {
    // Serialize writes using mutex pattern
    this.writeMutex = this.writeMutex.then(async () => {
      if (this.closed) {
        throw new Error("Connection closed");
      }

      try {
        const bytes = this.codec.marshalMessages(msgSet);
        // Append newline for NDJSON format
        const newline = new Uint8Array([10]); // '\n'
        const bytesWithNewline = new Uint8Array(bytes.length + 1);
        bytesWithNewline.set(bytes);
        bytesWithNewline.set(newline, bytes.length);
        await this.writer.write(bytesWithNewline);
      } catch (err) {
        this.connError = err as Error;
        this.close();
        throw err;
      }
    });

    return this.writeMutex;
  }

  /**
   * Read loop - continuously reads messages from the stream
   */
  private async readLoop(): Promise<void> {
    try {
      while (!this.closed) {
        const { value, done } = await this.reader.read();

        if (done) {
          break;
        }

        // Feed chunk to stream decoder to extract complete JSON values
        const jsonMessages = this.streamDecoder.decode(value);

        // Process each complete JSON message
        for (const jsonBytes of jsonMessages) {
          // Decode message set
          const msgSet = this.codec.unmarshalMessages(jsonBytes);

          // Determine if all messages are requests or responses
          const allRequests = msgSet.messages.every((msg) => "method" in msg);
          const allResponses = msgSet.messages.every((msg) => "result" in msg || "error" in msg);

          if (allRequests && msgSet.isBatch) {
            // Handle batch requests
            this.handleIncomingBatch(msgSet).catch((err) => {
              console.error("Error handling batch:", err);
            });
          } else if (allRequests) {
            // Handle single request
            this.handleIncomingRequest(msgSet.messages[0] as Request).catch((err) => {
              console.error("Error handling request:", err);
            });
          } else if (allResponses) {
            // Handle responses
            for (const msg of msgSet.messages) {
              this.handleIncomingResponse(msg as Response);
            }
          }
        }
      }
    } catch (err) {
      if (!this.closed) {
        this.connError = err as Error;
      }
    } finally {
      this.close();
    }
  }

  /**
   * Handle an incoming request from the remote peer
   */
  private async handleIncomingRequest(req: Request): Promise<void> {
    const resp = await this.handler.handleRequest(req);

    // Send response if not a notification
    if (resp) {
      const msgSet: MessageSet = {
        messages: [resp],
        isBatch: false,
      };
      await this.writeMessage(msgSet).catch(() => {
        // Ignore write errors for responses
      });
    }
  }

  /**
   * Handle an incoming batch of requests
   */
  private async handleIncomingBatch(msgSet: MessageSet): Promise<void> {
    const batch = toBatch(msgSet);
    const responses = await this.handler.handleBatch(batch);

    if (responses.length > 0) {
      const responseMsgSet = batchResponseToMessageSet(responses);
      await this.writeMessage(responseMsgSet).catch(() => {
        // Ignore write errors for responses
      });
    }
  }

  /**
   * Handle an incoming response to our request
   */
  private handleIncomingResponse(resp: Response): void {
    const pending = this.pendingRequests.get(resp.id);
    if (!pending) {
      return;
    }

    this.pendingRequests.delete(resp.id);

    if (resp.error) {
      pending.reject(RpcError.fromObject(resp.error));
    } else {
      // Track remote references
      this.trackRemoteReferences(resp.result);
      // Promote ReferenceType objects to Reference class instances
      const promoted = promoteReferences(resp.result);
      pending.resolve(promoted);
    }
  }

  /**
   * Track remote references in results
   */
  private trackRemoteReferences(value: unknown): void {
    if (value === null || value === undefined) {
      return;
    }

    if (isReference(value)) {
      this.session.addRemoteRef(value.$ref);
      return;
    }

    if (Array.isArray(value)) {
      for (const item of value) {
        this.trackRemoteReferences(item);
      }
      return;
    }

    if (typeof value === "object") {
      for (const val of Object.values(value)) {
        this.trackRemoteReferences(val);
      }
    }
  }
}
